File: app/csharp/AccountingSystem.Domain/Events/JournalEntryApprovedEvent.cs
namespace AccountingSystem.Domain.Events;

/// <summary>
/// 仕訳承認イベント（イベントソーシング用）
/// </summary>
public record JournalEntryApprovedEvent : IEventSourcedDomainEvent
{
    public required string JournalEntryId { get; init; }
    public required string ApprovedBy { get; init; }
    public required string ApprovalComment { get; init; }
    public required DateTime OccurredAt { get; init; }
    public required string UserId { get; init; }

    public string AggregateId => JournalEntryId;
    public string EventType => "JournalEntryApprovedEvent";
    public int EventVersion => 1;
}

File: app/ruby/accounting_system/app/events/domain_events/journal_entry_approved_event.rb
# frozen_string_literal: true

module DomainEvents
  # 仕訳承認イベント
  class JournalEntryApprovedEvent
    attr_reader :journal_entry_id, :approved_by, :approval_comment, :occurred_at

    def initialize(journal_entry_id:, approved_by:, approval_comment:, occurred_at:)
      @journal_entry_id = journal_entry_id
      @approved_by = approved_by
      @approval_comment = approval_comment
      @occurred_at = occurred_at
    end

    def to_h
      {
        journal_entry_id: @journal_entry_id,
        approved_by: @approved_by,
        approval_comment: @approval_comment,
        occurred_at: @occurred_at.iso8601
      }
    end

    def to_json(*_args)
      to_h.to_json
    end
  end
end
